### PR TITLE
Move sld state from URL to sessionstorage

### PR DESCRIPTION
**Pull request by archendel**
SLD state was stored in URL, we should now store it in session storage

### CODE REVIEW

**Review by sBouzols — CHANGES_REQUESTED** [review]
some comments will test

File: src/redux/reducer.js [hunk 1]
```diff
@@ -456,6 +467,85 @@ export const reducer = createReducer(initialState, {
...
+        const sldToPinToggleIndex = sldState.findIndex(
+            (sld) => sld.id === action.id
+        );
+        if (sldToPinToggleIndex >= 0) {
+            // when found, if was opened, it's now PINNED
+            const sldToPinState = sldState[sldToPinToggleIndex].state;
+            if (sldToPinState === ViewState.OPENED) {
+                sldState[sldToPinToggleIndex].state = ViewState.PINNED;
+            } else if (sldToPinState === ViewState.PINNED) {
+                // if sld is unpinned, the sld that had the state OPENED is now MINIMIZED
```
[the remark this answers is not in the document]
Comment (archendel) [reply src/redux/reducer.js:530]: it is the current behaviour

File: src/redux/reducer.js [hunk 2]
```diff
@@ -456,6 +467,85 @@ export const reducer = createReducer(initialState, {
...
+            return;
+        }
+
+        // in the other cases, we will open the targeted sld
+        // previously opened sld is now MINIMIZED
+        const previouslyOpenedSldIndex = sldState.findIndex(
+            (sld) => sld.state === ViewState.OPENED
+        );
+        if (previouslyOpenedSldIndex >= 0) {
+            sldState[previouslyOpenedSldIndex].state = ViewState.MINIMIZED;
```
[the remark this answers is not in the document]
Comment (archendel) [reply src/redux/reducer.js:492]: it is the current behaviour

File: src/components/diagrams/singleLineDiagram/utils.js [hunk 3]
```diff
@@ -5,95 +5,90 @@
...
+    togglePinSld,
+} from '../../../redux/actions';
+import { useSelector } from 'react-redux';
+import { useEffect } from 'react';
+import { syncSldStateWithSessionStorage } from '../../../redux/session-storage';
 
 export const ViewState = {
     PINNED: 'pinned',
     MINIMIZED: 'minimized',
+    OPENED: 'opened',
```
[the remark this answers is not in the document]
Comment (archendel) [reply src/components/diagrams/singleLineDiagram/utils.js:24]: will probably do in the future, in another PR

File: src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js [hunk 4]
```diff
@@ -292,69 +248,39 @@ export function SingleLineDiagramPane({
...
-                const oldState = oldValue.get(id);
-                if (oldState === state) newVal.delete(id);
-                else newVal.set(id, state);
-                if (state !== ViewState.MINIMIZED) {
-                    if (type === SvgType.VOLTAGE_LEVEL) openVoltageLevel(id);
-                    else if (type === SvgType.SUBSTATION) openSubstation(id);
-                }
-                return newVal;
-            });
+            if (state === ViewState.MINIMIZED) {
```
[the remark this answers is not in the document]
Comment (archendel) [reply src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js:257]: going to improve the code readability to make it clearer

File: src/redux/session-storage.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import { APP_NAME } from '../utils/config-params';
```
[the remark this answers is not in the document]
Comment (archendel) [reply src/redux/session-storage.js:1]: fixed

**Review by niconoir — APPROVED** [review]

**Review by Walid-Sahnoun — APPROVED** [review]

**Review by sBouzols — APPROVED** [review]

**Review by sBouzols — COMMENTED** [review]
some comments

**Review by sBouzols — APPROVED** [review]
Code Review ok (some ligth comments)
Tests done OK
* Duplicate tab on Firefox copy the content of sessionStorage, this is cool.
Console warning check OK

File: src/redux/session-storage.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        SESSION_STORAGE_SLD_STATE_KEY,
+        JSON.stringify({
+            studyUuid: studyUuid,
+            sldtate: sldState,
+        })
+    );
+};
+
+export const loadSldStateFromSessionStorage = (studyUuid) => {
+    const sldState = JSON.parse(
```
[the remark this answers is not in the document]
Comment (archendel) [reply src/redux/session-storage.js:30]: fixed

File: src/redux/session-storage.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        sessionStorage.getItem(SESSION_STORAGE_SLD_STATE_KEY)
+    );
+
+    if (sldState?.studyUuid === studyUuid) {
+        return sldState.sldtate;
```
[the remark this answers is not in the document]
Comment (archendel) [reply src/redux/session-storage.js]: fixed

File: src/redux/session-storage.js [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+export const syncSldStateWithSessionStorage = (sldState, studyUuid) => {
+    if (studyUuid == null) {
+        return;
+    }
+
+    sessionStorage.setItem(
+        SESSION_STORAGE_SLD_STATE_KEY,
+        JSON.stringify({
+            studyUuid: studyUuid,
+            sldtate: sldState,
```
[the remark this answers is not in the document]
Comment (archendel) [reply src/redux/session-storage.js]: fixed

File: src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js [hunk 9]
```diff
@@ -435,8 +339,9 @@ export function SingleLineDiagramPane({
                             showInSpreadsheet={showInSpreadsheet}
                             loadFlowStatus={loadFlowStatus}
                             numberToDisplay={displayedSLD.length}
-                            toggleState={toggleState}
-                            pinned={viewState.get(sld.id) === ViewState.PINNED}
+                            togglePin={togglePinSld}
+                            minimize={minimizeSld}
```
[the remark this answers is not in the document]
Comment (archendel) [reply src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js:343]: fixed

**Review by jonenst — APPROVED** [review]

**Pull request merged**